### PR TITLE
core: Add compile time byte swapping (C11 version)

### PR DESCRIPTION
**Pull request by Kijewski**
Please see #4812 first.

Using C11 you can tell the the value of an expression is known at compile:

``` c
_Generic(
    ((1 ? (char*) 0 : (void*) (ptrdiff_t) ((EXPR) - (EXPR)))),
    char*: COMPILE_TIME,
    default: RUN_TIME
)
```

This PR uses this feature to optimize HTONL and friends. They'll use shift-and-or if the value is known at compile time, otherwise `__builtin_bswap…()`.

Drawback: `_Generic` needs C11 and is only implemented since GCC 4.9.

### CODE REVIEW

**Comment by Kijewski** [comment]
Major tag because of `-std=c11`

**Comment by authmillenon** [comment]
I don't get what this macro does... @DipSwitch's code on the other hand I understand. Also what is with `clang` support?

**Comment by Kijewski** [comment]
Clang supports generic selections since somewhen in 2012. Source: some blog that comes up if you google "clang generic selection".

**Comment by Kijewski** [comment]
The macro "selects" either `byteorder_swaplc(V)` (shift-and-or) or `byteorder_swapl(V)` (compiler intrinsic), depending on whether `V` is known at compile time.

It uses the type deduction in:
```c
1 ? (char*) 0 : (void*) (ptrdiff_t) ((EXPR) - (EXPR))
```
If the value of `EXPR` is known at compile time, then the value of the whole ternary-expression is known at compile time. So the compiler can tell, that to value is a `char*`, because `1` evaluates to true. If `EXPR` is not know, then the compiler has to use the common type of `char*` and `void*`, i.e. `void*`. The generic selection either follows the `char*` or the default case, depending on the outcome. The other case is simply thrown away as if was `#if 0`'d (well, it has to use proper syntax).

**Comment by gebart** [comment]
The biggest issue is probably the msp430 platforms which are currently stuck at gcc 4.6 for most users. The other platforms will likely be at 4.9 or later anytime soon because of good upstream support for the ports.
To use the newer msp toolchain we also need to add newlib support to msp430 and do some changes to the startup code. I don't have any msp430 hardware but if anyone wants to do it you can adopt my PR #4766 which has most of the work done, just needs a rebase and some linking fixes after some code refactorings in master.

**Comment by Kijewski** [comment]
Hm, I thought a `HTONL` function that just does the optimal thing automagically would be the easiest option, but msp430 support is essential of course.

**Comment by DipSwitch** [comment]
Interesting :), I'll soon give it a try :)

**Comment by gebart** [comment]
Won't make it for release because of missing msp430 support

File: Makefile.cflags [hunk 1]
```diff
@@ -3,10 +3,10 @@
 # Otherwise test if the compiler unterstands the "-std=c99" flag, and use it if so.
 ifeq ($(filter -std=%,$(CFLAGS)),)
   ifeq ($(shell $(CC) -std=gnu99 -E - 2>/dev/null >/dev/null </dev/null ; echo $$?),0)
-    CFLAGS += -std=gnu99
+    CFLAGS += -std=gnu11
   else
     ifeq ($(shell $(CC) -std=c99 -E - 2>/dev/null >/dev/null </dev/null ; echo $$?),0)
-      CFLAGS += -std=c99
+      CFLAGS += -std=c11
```
Comment (miri64) [inline Makefile.cflags]: Errrrr... portability?

Comment (Kijewski) [inline Makefile.cflags]: Please read the comments in this PR ...

Comment (kaspar030) [inline Makefile.cflags]: Anyhow, this first checks for c99 and if successful, sets c11.

Comment (miri64) [inline Makefile.cflags]: Ah, sorry for the noise... But up until this is resolved I would prefer the portable version at #3812.

**Comment by miri64** [comment]
Okay it seems we progressed with c11 anyway, but the byteorder functions changed a little bit (they are now lower-case). @DipSwitch do you want to keep the upper-case variant for constant initialization (would be like we do with most other data types too). And after all: needs rebase.

**Comment by smlng** [comment]
needs rebase, but also discussion seems inconclusive, and was postponed several times now. Remove milestone, for time being.

**Comment by tcschmidt** [comment]
@miri64 do you see a chance to progress this? Or shall we close?

**Comment by miri64** [comment]
No real progress for >1 year, so I would say close. If someone disagrees, please don't hesitate reopen.

**Pull request closed without merging**